### PR TITLE
Provide a timeout function to fail quick on blocking operations

### PR DESCRIPTION
**Pull request by Huang-Wei**
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

Follow-up of https://github.com/kubernetes/kubernetes/pull/105016#discussion_r708558567.

When we run a blocking operation in the integration test, if the operation function well, the entire test would complete properly. But in terms of abnormal cases, like a PR introduces faulty changes, the blocking operation would hang there until the default go test timeout (10m) is reached.

```
$ go help testflag
...
	-timeout d
	    If a test binary runs longer than duration d, panic.
	    If d is 0, the timeout is disabled.
	    The default is 10 minutes (10m).
```

This PR introduces a timeout function to fail quickly in this scenario. It works similarly to the Linux `timeout` utility. 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: test/integration/scheduler/queue_test.go [hunk 1]
```diff
@@ -163,8 +171,14 @@ func TestServiceAffinityEnqueue(t *testing.T) {
 		t.Fatalf("Failed to delete service 'svc': %v", err)
 	}
 
-	// Now we should be able to pop the Pod from activeQ again.
-	podInfo = testCtx.Scheduler.NextPod()
+	// NextPod() is a blocking operation. Wrap it in timeout() to avoid relying on global integration
+	// test timeout (10m) to abort.
```
Comment (chendave) [inline test/integration/scheduler/queue_test.go:175]: I was trying to search where is the 10m defined, eventually, it seems like it is just a default setting for go testing. :)

>         -timeout d
            If a test binary runs longer than duration d, panic.
            If d is 0, the timeout is disabled.
            The default is 10 minutes (10m).


/lgtm
/hold
hold in case you want a double check from other reviewers, otherwise feel free to lift the label.

Comment (Huang-Wei) [reply test/integration/scheduler/queue_test.go:175]: Thanks for pointing it out. I will update.

**Comment by Huang-Wei** [comment]
/cc @alculquicondor 
to take another look.

File: test/integration/scheduler/queue_test.go [hunk 2]
```diff
@@ -303,8 +317,15 @@ func TestCustomResourceEnqueue(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Pop fake-pod out. It should be unschedulable.
-	podInfo := testCtx.Scheduler.NextPod()
+	var podInfo *framework.QueuedPodInfo
+	// NextPod() is a blocking operation. Wrap it in timeout() to avoid relying on global integration
+	// test timeout (10m) to abort.
+	if err := timeout(testCtx.Ctx, time.Second*15, func() {
```
Comment (alculquicondor) [inline test/integration/scheduler/queue_test.go:323]: Maybe all these lines could be a util function too. I would call it `nextPodorFatal` and accept `t` to be able to call Fatal from it.

Comment (Huang-Wei) [reply test/integration/scheduler/queue_test.go]: SG. `nextPodOrDie` may read better.

**Comment by Huang-Wei** [comment]
/label tide/merge-method-squash

@alculquicondor PTAL.

File: test/integration/scheduler/util.go [hunk 3]
```diff
@@ -511,3 +512,38 @@ func createNamespacesWithLabels(cs clientset.Interface, namespaces []string, lab
...
+	case <-done:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// nextPodOrDie returns the next Pod in the scheduler queue.
+// The operation needs to be completed within 15 seconds; otherwise the test gets aborted.
+func nextPodOrDie(t *testing.T, testCtx *testutils.TestContext) *framework.QueuedPodInfo {
```
Comment (chendave) [inline test/integration/scheduler/util.go:538]: we lost the flexibility to config how long the test case should abort, it's hard coded as 15 seconds now, is that necessary to pass another var here?

File: test/integration/scheduler/queue_test.go [hunk 4]
```diff
@@ -138,8 +138,7 @@ func TestServiceAffinityEnqueue(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Pop Pod2 out. It should be unschedulable.
```
Comment (chendave) [inline test/integration/scheduler/queue_test.go:141]: sound like those comments are still valid.

File: test/integration/scheduler/util.go (hunk 3 above)
Comment (Huang-Wei) [reply test/integration/scheduler/util.go:538]: Let's see if we have this requirement in the future. If so, we can easily compose another `nextPodOrDieWithTimeout` function.

File: test/integration/scheduler/queue_test.go (hunk 4 above)
Comment (Huang-Wei) [reply test/integration/scheduler/queue_test.go:141]: correct, revert those comments back.

File: test/integration/scheduler/util.go (hunk 3 above)
Comment (alculquicondor) [reply test/integration/scheduler/util.go:538]: Add `t.Helper`

File: test/integration/scheduler/util.go [hunk 5]
```diff
@@ -511,3 +512,38 @@ func createNamespacesWithLabels(cs clientset.Interface, namespaces []string, lab
...
+	var podInfo *framework.QueuedPodInfo
+	// NextPod() is a blocking operation. Wrap it in timeout() to avoid relying on
+	// default go testing timeout (10m) to abort.
+	if err := timeout(testCtx.Ctx, time.Second*15, func() {
+		// Now we should be able to pop the Pod from activeQ again.
```
Comment (alculquicondor) [inline test/integration/scheduler/util.go:543]: This comment seems irrelevant here.

File: test/integration/scheduler/util.go (hunk 3 above)
Comment (Huang-Wei) [reply test/integration/scheduler/util.go:538]: done.

File: test/integration/scheduler/util.go (hunk 5 above)
Comment (Huang-Wei) [reply test/integration/scheduler/util.go]: done.

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/hold cancel
